decidim-admin/app/forms/decidim/admin/permissions_form.rb

```diff
@@ -9,7 +9,12 @@ class PermissionsForm < Form
 
       attribute :permissions, Hash[String => PermissionForm]
 
-      def valid?
+      private
+
+      # Overriding Rectify::Form#form_attributes_valid? to preserve errors from custom method validations.
+      def form_attributes_valid?
```

Comment:
ActiveRecord allows to specify custom validations via methods, like `validates :check_permissions_are_valid` or similar.
Can't we do the same with FormObjects?

Reply:
The problem is that unless custom method validations add errors to the Form object itself (not to the attributes), those errors will be ignored.

Example:

```ruby
# decidim-admin/app/forms/decidim/admin/permissions_form.rb
def custom_validation
  errors.add(:base) # This will make Rectify::Form#valid? to fail
end

def custom_validation
  permissions["amend"].errors.add(:base, "ERROR") # This will be ignored
end
```
This is because when validating the attributes (after custom method validations) it does not check if the attributes already contain errors.

```ruby
# rectify-0.13.0/lib/rectify/form.rb
def valid?(options = {})
  before_validation

  options     = {} if options.blank?
  context     = options[:context]
  validations = [super(context)] # Here is where custom method validations are run

  validations << form_attributes_valid? unless options[:exclude_nested]
  validations << array_attributes_valid? unless options[:exclude_arrays]

  validations.all?
end
```

Reply:
I understand that, but can't we implement a custom validation method like 
```
def custom_validation
  errors.add(:base) # This will make Rectify::Form#valid? to fail
end
```

instead of overriding the parent method?

Reply:
Yes, but what we want is to be able to add errors to permissions related to a given action, which would look like this:

![action_permission_errors](https://user-images.githubusercontent.com/40306853/68148329-0f389d80-ff3c-11e9-8baf-7f5773c76167.png)


Also, the form was already overriding `valid?` because `Rectify::Form` does not validate hash attributes.